SimG4Core/Watcher/interface/SimWatcherMaker.h

```diff
@@ -34,24 +36,13 @@ class SimWatcherMaker : public SimWatcherMakerBase {
   SimWatcherMaker() {}
 
   // ---------- const member functions ---------------------
-  void make(const edm::ParameterSet &p,
-            SimActivityRegistry &reg,
-            std::shared_ptr<SimWatcher> &oWatcher,
-            std::shared_ptr<SimProducer> &oProd) const override {
-    auto returnValue = std::make_shared<T>(p);
-    SimActivityRegistryEnroller::enroll(reg, returnValue.get());
-    oWatcher = returnValue;
-
-    // If this is also a SimProducer, set the value
-    oProd = this->getSimProducer(returnValue.get(), returnValue);
-  }
-
-private:
-  std::shared_ptr<SimProducer> getSimProducer(SimProducer *, std::shared_ptr<T> &iProd) const {
-    return std::shared_ptr<SimProducer>(iProd);
-  }
-  std::shared_ptr<SimProducer> getSimProducer(void *, std::shared_ptr<T> &iProd) const {
-    return std::shared_ptr<SimProducer>();
+  void makeWatcher(const edm::ParameterSet& p,
+                   SimActivityRegistry& reg,
+                   SimWatcher* oWatcher,
+                   SimProducer* oProd) const override {
+    oWatcher = new T(p);
+    SimActivityRegistryEnroller::enroll(reg, oWatcher);
+    oProd = static_cast<SimProducer*>(oWatcher);
```

Comment:
Is this really safe? By quick look it seems to me that there are classes that inherit from `SimWatcher` but not `SimProducer`.

Reply:
SimProducer assumes to produce some data while SimWatcher do not, so SimProducers is a sub-group of SimWatchers. It is the same as before.

Reply:
Earlier (currently) `oProd == oWatcher` only if `T` inherits from `SimProducer`, and otherwise `oProd == nullptr`.

Reply:
@makortel, @bsunanda , I basically return the old signatures for making of SimWatcher. The only different - 2 new methods in SimWatcher, which allows to define consumes and tokens beginRun, so the module may run in the MT mode. Another methods set/get flag of MT, by default is false but may be changed by derived class.